### PR TITLE
[4.0] Media field, Media Manager fixes

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/29147 .

#### Media Manager: 
- remove the logic for dispatching event (used in the media field) from the item component to the store
- Add dispatch event for all the single select item operations
- Add dispatch event for the unselect item operation

#### Media field
- Remove couple of Joomla scoped variables/functions (not needed)
- Tweak the fetchImageDetails to reject immediately if there is no selected image in media manager
- Proper use of the promise when calling the fetchImageDetails fn

### Testing Instructions
Case 1: Normal operation
Create a new article and insert an image to the intro image.
Reset the field by pressing the remove button

Case 2: Selection problem
Create an article and go to the intro image opening the modal for the Media manager. Select an image by clicking on it. Now click anywhere in the white space below the list of images. The image should be unselected! Clicking the select button at the end of the modal should:
- Not insert anything in the field (this was previously returning the last image selected)
- Close the modal (same behaviour as 3.x)

Case 3: Race Condition
I can safely replicate this by adding a breakpoint and using debugger in the PHP world.
Basically what happens is that selecting an image is not synchronous so if you're too fast to select an image and then click on the select button on the bottom of the modal you will not get the selected image path in the input field. Probably can be reproduced with the dropbox plugin(?)

### Expected result
All cases should work

### Documentation Changes Required

Nope

### CODE REVIEW

File: build/media_source/system/js/fields/joomla-field-media.w-c.es6.js [hunk 1]
```diff
@@ -192,9 +195,10 @@
 
     modalClose() {
       const input = this.querySelector(this.input);
-      Joomla.getImage(Joomla.selectedFile, input, this);
 
-      Joomla.Modal.getCurrent().close();
+      fetchImageDetails(selectedFile, input, this)
+        .then(() => { Joomla.Modal.getCurrent().close(); })
+        .catch(() => { Joomla.Modal.getCurrent().close(); });
```
Comment (wilsonge) [inline build/media_source/system/js/fields/joomla-field-media.w-c.es6.js]: If this fails we still should probably show an error to the user otherwise the modal closes but potentially thing updates etc.

File: administrator/components/com_media/resources/scripts/store/mutations.js [hunk 2]
```diff
@@ -229,6 +229,17 @@ export default {
      */
     [types.SELECT_BROWSER_ITEM]: (state, payload) => {
         state.selectedItems.push(payload);
+
+        window.parent.document.dispatchEvent(new CustomEvent('onMediaFileSelected', {
```
Comment (wilsonge) [inline administrator/components/com_media/resources/scripts/store/mutations.js]: We have to be careful here? previously only files threw this event - now it's any item. First of all the event being thrown isn't quite correct now. But it also means with this PR you can probably select a folder in the media modal

File: build/media_source/system/js/fields/joomla-field-media.w-c.es6.js (hunk 1 above)
Comment (dgrammatiko) [reply build/media_source/system/js/fields/joomla-field-media.w-c.es6.js]: > If this fails

Probably we need to throw an alert for any possible server bad response (403, 500, etc)

File: administrator/components/com_media/resources/scripts/store/mutations.js (hunk 2 above)
Comment (dgrammatiko) [reply administrator/components/com_media/resources/scripts/store/mutations.js]: Pfff, true I'll move the select part back to the item and figure a way to safely throw the unselect part

File: build/media_source/system/js/fields/joomla-field-media.w-c.es6.js (hunk 1 above)
Comment (dgrammatiko) [reply build/media_source/system/js/fields/joomla-field-media.w-c.es6.js]: Are you ok with a simple alert like the one in your PR for anything that will fall in the catch?

**Comment by wilsonge** [comment]
I have tested this item :white_check_mark: successfully on 5c6a395d801a6509ec9db6826b65ed5536afe7a5<br /><br />This works and is better than before. However selecting no Image does give the error message "Internal server error." still although obviously displaying this visibly is an improvement on the silent fail before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29256">issues.joomla.org/tracker/joomla-cms/29256</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!
